### DISCUSSION TITLE
Tool-bar plugin icons not scaling

### DISCUSSION THREAD

**Discussion by croaker000** [pulsar-edit/.github#201] [Community Packages]
Pulsar 1.107.1, Tool-bar 1.4.4 plugin

With this release, the toolbar icons no longer scale when the toolbar icon size is changed. The tool-bar width does change, the icons stay the same size. Was fine in 1.106

**Comment by Daeraxa** [comment]
This is probably down to the `less` upgrade and the requirement for `parens-division`, there are lines like this:
https://github.com/atom-community/tool-bar/blob/6ec4827551ef374fd796787d7283d001d38b84d4/styles/tool-bar.less#L145C11-L145C11
Which I'm guessing are causing the issue, as far as I know Pulsar should be trying to fix this on the fly - @confused-Techie do we need this as a bug or is this being tracked already?

**Reply by confused-Techie** [reply]
This is not being tracked, but I don't think we need a bug. 

Unless this usage is coming from a bundled package, Pulsar should try to upgrade it automatically. Some users have reported they have to relaunch Pulsar a second time for the fixes to happen. But I'm assuming the package's Less is still in the cache. If Pulsar isn't automatically fixing a community package's less I'd recommend deleting all of the content in `HOME/.pulsar/compile-cache/less` and relaunching.

That is unless of course, this package is being used within one of the bundled package's in Pulsar, in which case it will not be automatically upgraded

**Reply by croaker000** [reply]
A plain restart makes no different. Closing, clearing the less cache and restarting makes no difference.

**Reply by confused-Techie** [reply]
@croaker000 Since the `tool-bar` package doesn't seem to do much on it's own, it looks like it's intended to be consumed by other packages. Can you provide a package that does consume this?  Or could you provide a the code for how you are modifying this? So that I can attempt to replicate the issue.

But I appreciate you trying to clear the cache, so this either means something else is going on here, as in not because of the Less update, or for some reason this usage isn't being caught by the automatic upgrade

**Reply by confused-Techie** [reply]
Actually wait, looking through the source, the lines @Daeraxa pointed out wouldn't be automatically upgraded. It's failing to be matched by the automatic upgrades. (I just tested those lines manually).

So that very likely is what's causing the issue. I'll go ahead and create a proper issue for this so we can track it

**Reply by croaker000** [reply]
Tool-bar works in conjunction with tool-bar-main which is basically a tiny package that contains a sample definition of toolbar icons that can be changed by editing the configuration file.

**Reply by confused-Techie** [reply]
@croaker000 Thanks for the usage information.

I've gone ahead and created an [issue](https://github.com/pulsar-edit/pulsar/issues/650) for this, so we can track it there. Where I'd also recommend to check in for further developments. Thanks a ton for reporting this

**Reply by croaker000** [reply]
I notice the issue marked "resolved" and I've now updated to 1.108 and still have the problem. Is there anything I need to do my end or has the fix not made it to distribution yet?

**Reply by confused-Techie** [reply]
> I notice the issue marked "resolved" and I've now updated to 1.108 and still have the problem. Is there anything I need to do my end or has the fix not made it to distribution yet?

The changes I made intended to fix this issue are currently in this release of Pulsar. Ill have to investigate to see if theres something else going on causing the issue to still occur.

Although something you can try in the meantime, is clear the package-cache folder within you `~/.pulsar` folder, to see if it resolves it

**Reply by croaker000** [reply]
I don't have a package-cache. Is that a bad thing?

**Reply by confused-Techie** [reply]
@croaker000 Sorry, was speaking from memory to early in the morning for me.

You'll want to clear `~/.pulsar/compile-cache/less`, this is the folder that contains all compiled CSS, so could be keeping the error around, instead of retranspiling it.

**Reply by croaker000** [reply]
Thanks. No, still the same :(

**Reply by confused-Techie** [reply]
So looking at this again, seems there's a new line causing the failure.

[Here](https://github.com/atom-community/tool-bar/blob/6ec4827551ef374fd796787d7283d001d38b84d4/styles/tool-bar.less#L97) we can see an interesting construct:

```less
&.tool-bar-24px {
  .tool-bar-size(24px * 2);
}
```

Now, I'd have to double check, but if memory serves, this won't be caught by the Less compatibility "Transform Deprecated Math Usage in Less Stylesheets" setting, because it's already enclosed in parenthesis. Which when I drafted that feature up, I had thought anything in parenthesis would be handled by less automatically.

But since this is a call to another function of this Less style sheet, I suppose it's not. Since this snippet results in the following CSS:

```css
.tool-bar.tool-bar-24px .tool-bar-btn:before {
  font-size: 48px / 2;
  line-height: 48px;
}
```

Which, is obviously invalid. The hard part is I'm not sure what's best to do about this, because we can't have the "Transform Deprecated Math Usage" effect everything, even if in parenthesis', since that would begin to break some functions, I'd imagine, the biggest concern being that it would match on things it's already upgraded, causing the time to transform a single file would increase.

So yeah, this one is a little hard to say. The best thing to do, may be just to actually fix the less code itself, unless someone can think of how we can force this package to work on newer versions of less.

Although, I will say, this package is from `atom-community` we could look into forking it and having it official supported under Pulsar.

**Reply by confused-Techie** [reply]
Must have looked at this too fast.

Looking again, this is in fact the [line causing the error](https://github.com/atom-community/tool-bar/blob/6ec4827551ef374fd796787d7283d001d38b84d4/styles/tool-bar.less#L149).

Which makes much more sense when considering the final CSS that's output. Although, this should absolutely be caught as something that should be fixed.

So once I'm able I'll try to run some additional tests

**Reply by croaker000** [reply]
Not sure if this helps but changing line 145 of [tool-bar.less](https://github.com/atom-community/tool-bar/blob/6ec4827551ef374fd796787d7283d001d38b84d4/styles/tool-bar.less#L145) to

`font-size: @size * 0.5;`

Seems to have done the trick.

**Reply by gabriel-xlab** [reply]
Following @croaker000, for text based buttons you should also change for the remaining lines:

```css
    .tool-bar-btn {
      height: @size + (@button-border-size * 2);
      min-width: @size + (@button-border-size * 2);
      &:before {
        font-size: @size * 0.5;
        line-height: @size;
      }

      &:not(:empty) {
        font-size: @size * 0.5;
        line-height: @size;
        padding: 0 @size * 0.25;
        &:before {
          padding-right: @size * 0.25;
        }
      }
```

**Reply by thatcomputerguy0101** [reply]
I notice there is now [a fork of this package](https://github.com/pulsar-edit/tool-bar) under the pulsar-edit organization, but with no commits beyond the original repo. Is it being taken on as a pulsar package? I would be willing to create a PR somewhere to help fix this, but I'm unsure where I should do so.
